### PR TITLE
Added support for sparse tensors to PyTorch DistributedOptimizer

### PR DESCRIPTION
**Pull request by tgaddair**
By default, sparse gradients will be aggregated using `allgather` similar to how this feature works for TensorFlow.

Alternatively, the user can set `hvd.DistributedOptimizer(..., sparse_as_dense=True)`, converting sparse gradients to dense to force usage of allreduce instead of allgather.  This can improve performance if gradients do not have many missing values.

### CODE REVIEW

File: horovod/torch/optimizer.py [hunk 1]
```diff
@@ -219,9 +241,13 @@ def synchronize(self):
                     self._allreduce_delay[gp] = self.backward_passes_per_step
                     gp.grad.set_(self._compression.decompress(output, gctx))
             else:
-                output = synchronize(handle)
+                # When handle is a callable function, it returns the aggregated tensor result
+                output = synchronize(handle) if not callable(handle) else handle()
                 self._allreduce_delay[p] = self.backward_passes_per_step
-                p.grad.set_(self._compression.decompress(output, ctx))
+                if p.grad.is_sparse:
+                    p.grad.zero_().add_(self._compression.decompress(output, ctx))
```
Comment (chongxiaoc) [inline horovod/torch/optimizer.py]: I see, put sparse result to original dense zero matrix.

Comment (tgaddair) [reply horovod/torch/optimizer.py]: For some reason you can't `set_` a sparse tensor, so we need to use this hack.  It's kind of ugly.

**Review by chongxiaoc — APPROVED** [review]
look good to me

**Review by romerojosh — APPROVED** [review]
LGTM!
